clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -341,30 +341,45 @@ static void collectSYCLAttributes(Sema &S, FunctionDecl *FD,
       FD->dropAttr<SYCLIntelUseStallEnableClustersAttr>();
     }
   }
+
+  // Attributes that should not be propagated from device functions to a kernel.
+  if (DirectlyCalled) {
+    llvm::copy_if(FD->getAttrs(), std::back_inserter(Attrs), [](Attr *A) {
+      return isa<SYCLIntelLoopFuseAttr, SYCLIntelFPGAMaxConcurrencyAttr,
+                 SYCLIntelFPGADisableLoopPipeliningAttr,
+                 SYCLIntelFPGAInitiationIntervalAttr>(A);
+    });
+  }
 }
 
-class MarkDeviceFunction : public RecursiveASTVisitor<MarkDeviceFunction> {
+class DiagDeviceFunction : public RecursiveASTVisitor<DiagDeviceFunction> {
   // Used to keep track of the constexpr depth, so we know whether to skip
   // diagnostics.
   unsigned ConstexprDepth = 0;
+  Sema &SemaRef;
+  const llvm::SmallPtrSetImpl<FunctionDecl *> &RecursiveFuncs;
+
   struct ConstexprDepthRAII {
-    MarkDeviceFunction &MDF;
+    DiagDeviceFunction &DDF;
     bool Increment;
 
-    ConstexprDepthRAII(MarkDeviceFunction &MDF, bool Increment = true)
-        : MDF(MDF), Increment(Increment) {
+    ConstexprDepthRAII(DiagDeviceFunction &DDF, bool Increment = true)
+        : DDF(DDF), Increment(Increment) {
       if (Increment)
-        ++MDF.ConstexprDepth;
+        ++DDF.ConstexprDepth;
     }
     ~ConstexprDepthRAII() {
       if (Increment)
-        --MDF.ConstexprDepth;
+        --DDF.ConstexprDepth;
     }
   };
 
 public:
-  MarkDeviceFunction(Sema &S)
-      : RecursiveASTVisitor<MarkDeviceFunction>(), SemaRef(S) {}
+  DiagDeviceFunction(
+      Sema &S, const llvm::SmallPtrSetImpl<FunctionDecl *> &RecursiveFuncs)
```

Comment:
I was hoping we could make `RecursiveFuncs` hold `const FunctionDecl *` because nothing mutates the functions once they get passed in here. However, I see why that's really annoying (because we take the container rather than using a more general interface with iterators), so I suppose this is as const correct as we're likely to get.

Reply:
Actually, DeviceFunctions cannot be const I think, but the RecursiveFuncs I can make const.  I'll do that, thanks!